Add blog post about micro frontends for Java microservices (#1263)
oktadeveloper/okta-blog: _source/_posts/2022-09-26-micro-frontends-java-microservices.adoc

---
layout: blog_post
title: Micro Frontends for Java Microservices
author: matt-raible
by: advocate
communities: [java,javascript]

> description: ""

Reviewer: missing description

> description: ""

Author: ```suggestion
description: "This tutorial explains what micro frontends are and how they can be used with Java microservices."
```

I'm going to use "micro frontends" for the remainder of this post since that's what https://twitter.com/thecamjackson[Cam Jackson] used in his https://martinfowler.com/articles/micro-frontends.html[Micro Frontends article] on Martin Fowler's blog.

== A quick introduction to Module Federation

Webpack's https://webpack.js.org/concepts/module-federation/[Module Federation] is one of the best known implementations for micro frontends. Its https://webpack.js.org/plugins/module-federation-plugin[`ModuleFederationPlugin`] allows a build to provide or consume modules with other independent builds at runtime. It even allows you to share libraries between frontends to reduce the size of remote bundles.

https://twitter.com/ScriptedAlchemy[Zack Jackson] is the creator of Module Federation and recently https://twitter.com/ScriptedAlchemy/status/1564411584851505153[collaborated] with https://twitter.com/ManfredSteyer[Manfred Steyer] to create Native Federation. This means you can use micro frontend concepts with any build tool, not just webpack.

// https://twitter.com/ManfredSteyer/status/1564312149580582912
++++
{% twitter 1564312149580582912 %}
++++

== Why should Java developers care?

> I think micro frontends an excellent architecture concept. With microservices, you weave everything together on the backend with protocols like HTTP, gRPC, and GraphQL. With micro frontends, it's all HTTP. You can literally see your app get stitched together by watching your browser's network console and seeing remote modules load.

Reviewer: the first line reads a bit weird, but hey i'm not a native english speaker :P

Author: I updated it. Please have another look.

> I think micro frontends an excellent architecture concept. With microservices, you weave everything together on the backend with protocols like HTTP, gRPC, and GraphQL. With micro frontends, it's all HTTP. You can literally see your app get stitched together by watching your browser's network console and seeing remote modules load.

Reviewer: I wouldn't call GraphQL a protocol. Isn't it a query language and more like REST rather than HTTP or gRPC?

Author: Fixed.

> I think micro frontends an excellent architecture concept. With microservices, you weave everything together on the backend with protocols like HTTP, gRPC, and GraphQL. With micro frontends, it's all HTTP. You can literally see your app get stitched together by watching your browser's network console and seeing remote modules load.

Author: ```suggestion
I think micro frontends are a very interesting architectural concept. With microservices, you weave everything together on the backend with protocols like HTTP, gRPC, and GraphQL. With micro frontends, it's all HTTP. You can literally see your app get stitched together by watching your browser's network console and seeing remote modules load.
```

I've encountered quite a few monolith UIs in my time as a consultant. The backend was a beautiful microservice architecture, but it was all coupled together on the frontend. There's a good chance a lot of Java developers don't care about the UI because they just work on the beautiful backends. However, if you consider yourself a Java _web_ developer, micro frontends are as revolutionary as HTML5!

And that's the beauty of this tutorial, you don't have to write any micro frontends. JHipster can create them for you!

== Micro frontends in action with JHipster

I used a JDL (JHipster Domain Language) file named `reactive-ms.jdl` to create the reactive Java microservices tutorial I mentioned earlier. You can see this file online in the https://github.com/jhipster/jdl-samples/blob/main/reactive-ms.jdl[JDL samples repository].

> I copied this file to `reactive-mf.jdl` and changed a few things for this tutorial:

Reviewer: may be worth mentioning the `jhipster download reactive-ms.jdl` command here as well?

Author: I mention this below after they've created the `micro-frontends-jhipster` directory.

image::{% asset_path 'blog/micro-frontends-java/jdl-with-kubernetes.png' %}[alt=Additional K8s deployment,width=800,align=center,link={% asset_path 'blog/micro-frontends-java/jdl-with-kubernetes.png' %}]

TIP: The https://plugins.jetbrains.com/plugin/19697-jhipster-jdl[JHipster JDL Plugin] is a handy tool for working with JDL files.

=== Micro frontend options: Angular, React, and Vue

JHipster has support for the big three JavaScript frameworks: Angular, React, and Vue. All are implemented using TypeScript and a newly generated app should have around 70% code coverage, both on the backend and frontend.

There is also a https://github.com/jhipster/generator-jhipster-svelte[Svelte blueprint], but it does not support micro frontends at the time of this writing.

=== Build Java microservices with Spring Boot and WebFlux

To generate a microservices architecture with micro frontend support, run the following command:

[source,shell]
----
jhipster jdl reactive-mf.jdl --monorepository --workspaces
----

> The last two arguments are optional, but I expect you to use them for this tutorial. Without the `monorepository` flag, the gateway and microservices would have their own Git repos. The `workspaces` flag enables https://docs.npmjs.com/cli/v8/using-npm/workspaces[npm workspaces], which are kinda like having a parent `pom.xml` that allows you to execute commands across projects. It also makes it so there's only one `node_modules` in the root directory. To learn more, I recommend egghead's https://egghead.io/courses/introduction-to-monorepos-with-npm-workspaces-c03f500b[Introduction to Monorepos with NPM Workspaces].

Author: ```suggestion
The last two arguments are optional, but I expect you to use them for this tutorial. Without the `monorepository` flag, the gateway and microservices would have their own Git repos. The `workspaces` flag enables https://docs.npmjs.com/cli/v8/using-npm/workspaces[npm workspaces], which are kinda like having an aggregator `pom.xml` that allows you to execute commands across projects. It also makes it so there's only one `node_modules` in the root directory. To learn more, I recommend egghead's https://egghead.io/courses/introduction-to-monorepos-with-npm-workspaces-c03f500b[Introduction to Monorepos with NPM Workspaces].
```

> If you want to use Angular, you can pass in `--client-framework angularX` to override the JDL value. For example:

Reviewer: I think just `angular` should also work already as we list that as acceptable in the docs

Author: Nope. There's an error when I run it:

```
✖ An error occured while running jhipster:client#microfrontend
ERROR! Client framework angular doesn't support microfrontends
```

_Make sure your new user's email is verified before attempting to log in!_

Next, head to **Actions** > **Flows** and select **Login**. Create a new action named `Add Roles` and use the default trigger and runtime. Change the `onExecutePostLogin` handler to be as follows:

[source,js]
----
exports.onExecutePostLogin = async (event, api) => {
  const namespace = 'https://www.jhipster.tech';
  if (event.authorization) {
    api.idToken.setCustomClaim('preferred_username', event.user.email);
    api.idToken.setCustomClaim(`${namespace}/roles`, event.authorization.roles);
    api.accessToken.setCustomClaim(`${namespace}/roles`, event.authorization.roles);
  }
}
----

This code adds the user's roles to a custom claim (prefixed with `\https://www.jhipster.tech/roles`). This claim is mapped to Spring Security authorities in `SecurityUtils.java` in the gateway app.

Select **Deploy** and drag the `Add Roles` action to your Login flow.

Edit `docker-compose/central-server-config/application.yml` and append the following YAML block to add your Auth0 settings.

[source,yaml]
----
jhipster:
  security:
    ...
    oauth2:

>       audience: https://dev-06bzs1cu.us.auth0.com/api/v2/

Author: ```suggestion
      audience: https://<your-auth0-domain>/api/v2/
```

Author: Replace value with placeholder
